### ISSUE TITLE
Caching DescribeImages results (amiSelectorTerms)

### ISSUE THREAD

**Issue by ilkinmammadzada** [aws/karpenter-provider-aws#5997]
### Description

**What problem are you trying to solve?**
We have multiple NodePools/NodeClasses and but amiSelectorTerms field is identical for them. I'm seeing that karpenter sends `DescribeImages` requests for them separately and it causes `Client.RequestLimitExceeded` errors:

"requestParameters": {"nextToken": "x", "maxResults": 500, "executableBySet": {}, "imagesSet": {}, "ownersSet": {"items": [{"owner": "self"}, {"owner": "amazon"}]}, "filterSet": {"items": [{"name": "name", "valueSet": {"items": [{"value": "k8s-ami_*"}]}}, {"name": "tag:active", "valueSet": {"items": [{"value": "true"}]}}]}}

**How important is this feature to you?**
We are getting 3k+ errors in an hour. It isn't super critical but would be nice to solve. 

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by jonathan-innis** [comment]
Which version of Karpenter are you running on? Is this the latest version?

**Comment by ilkinmammadzada** [comment]
We recently upgraded to the v0.36 but it the issue still exist. 
I filtered all `RequestLimitExceeded` errors and seeing eventname is always `DescribeImages`.

~20 NodePools and nothing special.
And We have ~20 NodeClasses and all they have identical `amiSelectorTerms`:

```
amiSelectorTerms:
  - name: k8s-xxx-ami_*
    tags:
      active: "true"
```
I'm seeing that karpenter sends ~170 DescribeImages requests in a second (2024-04-17T09:17:44Z UTC) for the single k8s cluster. `requestParameters` is identical for all of them.

**Comment by jonathan-innis** [comment]
> I'm seeing that karpenter sends ~170 DescribeImages requests in a second

Do you see this consistently? Did this happen right after you did an upgrade or the Karpenter pod restarted and then consistently every 5m or so or does it happen intermittently?

**Comment by ilkinmammadzada** [comment]
Yes, I'm seeing consistently. No, it's not related upgrade or restart. It existed before the upgrade. 
I filtered single cluster and saw it happens every 5m.

**Comment by engedaam** [comment]
> I filtered single cluster and saw it happens every 5m.

This would seem related to the NodeClass status controller https://github.com/aws/karpenter-provider-aws/blob/bc653f0e1e09589e9b8c712a221ae64f6cb69b55/pkg/controllers/nodeclass/status/ami.go#L34. We are planning on using having the AMI data get pulled asynchronously to allow Karpenter to be more resilient and should elevate this issue.

**Comment by jonathan-innis** [comment]
> I'm seeing that karpenter sends ~170 DescribeImages requests in a second

@ilkinmammadzada If you were to perform the describe operation yourself, how many images would your EC2NodeClass match against?

**Comment by engedaam** [comment]
Just an update, there seems to be two causes for the high number of high requests you are experiencing.  
-  Karpenter currently does not have any locking mechanism during its read and write operation for AMIs that are stored in the cache. (This was the first guess outlined by [@jonathan-innis](https://github.com/jonathan-innis)). Solution will be adding in locks.
- The DescribeImages API does not filter out AMIs before fetching AMIs by tags. This will result in the DescribeImages API sending empty pages(This is currently expected behavior). The best mitigation here will for Karpenter to increases page sizes to reduce the number of request that are made.

Lastly, are using AMIs that are created by your team account? You might be able to reduce the number of requests today by specifying the owner field to being self or the account which the AMIs are owned.

**Comment by ilkinmammadzada** [comment]
Thanks for the update @engedaam !

Yes, we are using our own AMIs. Do we just need to add `owner: self` ? 

```
amiSelectorTerms:
  - name: k8s-xyz-ami_*
    **owner: self**
    tags:
      active: "true"
```

**Comment by engedaam** [comment]
> Do we just need to add owner: self ?

Yes!  I was able to reduce the number of request by adding the owner

**Comment by ilkinmammadzada** [comment]
@engedaam it helped reduce the number of requests significantly. Thanks!

**Comment by rtomadpg** [comment]
Slightly related and adding it here to help other Karpenter users optimise their setup and not hit any RequestLimitExceeded errors on the ec2:DecribeImages calls by Karpenter.

We had an `amiSelectorTerms` matching AMIs by name, without regex:
```
        amiSelectorTerms:
          - name: amazon-eks-node-1.28-v20240424 # ami-02876d93345f5c282
```

When we replaced the `name` term with an `id` term, like this:

```
        amiSelectorTerms:
          - id: ami-02876d93345f5c282 # amazon-eks-node-1.28-v20240424
```

... the ec2:DescribeImages call volume drop to near zero:


<img width="1428" alt="image" src="https://github.com/aws/karpenter-provider-aws/assets/151731240/7d1efd29-c128-479c-83ab-58937fd2480a">

Hope this helps.

Maybe the docs https://karpenter.sh/docs/concepts/nodeclasses/#specamiselectorterms should be updated to recommend using the default or `id` terms?
